File: 3867-sum-of-gcd-of-formed-pairs/3867-sum-of-gcd-of-formed-pairs.ts
function gcdSum(nums: number[]): number {
    const n = nums.length;
    const a: number[] = new Array(n);
    let mx = 0;

    for (let i = 0; i < n; i++) {
        mx = Math.max(mx, nums[i]);
        a[i] = gcd(nums[i], mx);
    }

    a.sort((x, y) => x - y);

    let ans = 0;
    let l = 0, r = n - 1;
    
    while (l < r) {
        ans += gcd(a[l], a[r]);
        l++;
        r--;
    }

    return ans;
}

// GCD
function gcd(a: number, b: number): number {
    while (b) { [a, b] = [b, a % b]; }
    return a;
}

File: 3867-sum-of-gcd-of-formed-pairs/3867-sum-of-gcd-of-formed-pairs.py
from math import gcd
from typing import List


class Solution:
    def gcdSum(self, nums: List[int]) -> int:
        n = len(nums)
        a = [0] * n
        mx = 0

        for i in range(n):
            mx = max(mx, nums[i])
            a[i] = gcd(nums[i], mx)

        a.sort()

        ans = 0
        l, r = 0, n - 1
        
        while l < r:
            ans += gcd(a[l], a[r])
            l += 1
            r -= 1

        return ans
